File: tree/257.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
import (
    "strconv"
)
func binaryTreePaths(root *TreeNode) []string {
    if root == nil {
        return nil
    }
    ret := make([]string, 0)
    var str string
    dfs(root, &ret, str)
    return ret
}

func dfs(root *TreeNode, ret *[]string, str string) {
    var newStr string
    if str == "" {
        newStr = strconv.Itoa(root.Val)
    }else {
        newStr = str + "->" + strconv.Itoa(root.Val)
    }
    if root.Left == nil && root.Right == nil {
        *ret = append(*ret, newStr)    
    }
    if root.Left != nil {
        dfs(root.Left, ret, newStr)
    }
    if root.Right != nil {
        dfs(root.Right, ret, newStr)
    }
}

File: tree/257.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        std::vector<std::string> ret;
        dfs(root, "", ret);
        return ret;
    }
private:
    void dfs(TreeNode* root, std::string cur, std::vector<std::string>& ret) {
        if (!root)
            return;
        if (!root->left && !root->right) {
            ret.emplace_back(cur+ std::to_string(root->val));
            return;
        }
        dfs(root->left, cur + std::to_string(root->val) + "->", ret);
        dfs(root->right, cur + std::to_string(root->val) + "->", ret);
        return;
    }
};
